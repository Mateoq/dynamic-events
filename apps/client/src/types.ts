export enum Direction {
  LEFT = 'left',
  RIGHT = 'right'
}

export enum CalendarType {
  MOTH = 'MONTH',
  WEEK = 'WEEK',
  DAY = 'DAY'
}

export enum ToastType {
  SUCCESS = 'success',
  ERROR = 'error'
}

export type Time = {
  key: string;
  hour: number;
  minutes: string[];
};

export type Day = {
  key: string;
  value: number;
  isToday: boolean;
  times: Record<string, Time>;
};

export type Week = {
  key: string;
  value: number;
  firstIndex: number;
  lastIndex: number;
  days: Record<string, Day>;
};

export type Month = {
  key: string;
  value: number;
  firstIndex: number;
  lastIndex: number;
  weeks: Record<string, Week>;
};

export type Year = {
  key: string;
  value: number;
  months: Record<string, Month>;
};

export type CalendarData = {
  years: Record<string, Year>;
};

export type DayToRender = {
  data: Day;
  week: number;
};

export type CalendarState = {
  year: number;
  month: number;
  week: number;
  day: number;
};

export type Event = {
  id: string;
  title: string;
  description: string;
  city: string;
  startYear: number;
  startMonth: number;
  startDate: number;
  endYear: number;
  endMonth: number;
  endDate: number;
  time: number;
  fullDay: boolean;
  color: string;
};

export type CreateEventDto = {
  title: string;
  description: string;
  city: string;
  startYear: number;
  startMonth: number;
  startDate: number;
  endYear: number;
  endMonth: number;
  endDate: number;
  time: number;
  fullDay: boolean;
  userId: string;
};

export type UpdateEventDto = Partial<CreateEventDto>;

export type CreateEventApiDto = CreateEventDto & {
  color: string;
};

export enum TransactionResultType {
  SUCCESS = 'TRAN_SUCCESS',
  EMAIL_USED = 'TRAN_EMAIL_USED',
  WRONG_ID = 'TRAN_WRONG_ID',
  WRONG_USER = 'TRAN_WRONG_USER',
  ERROR = 'TRAN_ERROR',
  UNKNOWN_ERROR = 'TRAN_UNKNOWN_ERROR'
}

export interface TransactionError {
  message: string;
}

export type TransactionResult<T> = {
  status: boolean;
  type: TransactionResultType;
  data: T;
  error: TransactionError | null;
};

export type EventMap = Record<string, Event>;

export type TimeEventMap = Record<number, string[]>;

export type DayEventMap = Record<number, TimeEventMap>;

export type MonthEventMap = Record<number, DayEventMap>;

export type YearEventMap = Record<number, MonthEventMap>;

export type UserEventMap = YearEventMap;

export type EventData = {
  events: EventMap;
  calendarEvents: UserEventMap;
};
// --------------------------------

export type DateItem = {
  key: string;
  date: Date;
  isToday: boolean;
};

export type DateMap = {
  [key: string]: DateItem;
};

export type MonthData = {
  firstIndex: number;
  lastIndex: number;
  dates: DateMap;
};

export type MonthMap = {
  [key: string]: MonthData;
};

export type CalendarMonthData = {
  currentKey: string;
  months: MonthMap;
  currentDate: {
    year: number;
    month: number;
  };
};
