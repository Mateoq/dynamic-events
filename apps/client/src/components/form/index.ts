export * from './Button';
export * from './SelectInput';
export * from './TextInput';
export * from './DatePicker';
export * from './CheckBox';
