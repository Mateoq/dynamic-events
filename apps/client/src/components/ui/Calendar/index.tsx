import React from 'react';

import { CalendarMonth } from './CalendarMonth';
import { CalendarWeek } from './CalendarWeek';
import { CalendarDay } from './CalendarDay';
import { CalendarNav } from './CalendarNav';
import { useCalendarStateStore } from '@/stores/calendar-state-store';
import { CalendarData, Direction, CalendarType } from '@/types';
import {
  getYearKey,
  getMonthKey,
  getWeekKey,
  getDayKey,
  getMonthFromState,
  getWeekFromState,
  getDayFromState
} from '@/utils/calendar';

export type CalendarProps = {
  type: CalendarType;
  data: CalendarData;
  onNav: (direction: Direction) => void;
  onCreateEvent: () => void;
  onChangeView: (type: CalendarType) => void;
  onClickEvent: (id: string) => void;
};

export const Calendar: React.FC<CalendarProps> = (props) => {
  const { type, data, onNav, onCreateEvent, onChangeView, onClickEvent } =
    props;
  const state = useCalendarStateStore((state) => state.calendarState);
  let currentComponent = (
    <div className="text-red-600 text-xl">Unknown Calendar</div>
  );

  console.log('DATA', data);

  if (state.year === 0 || state.month === 0 || state.day === 0) {
    return null;
  }

  switch (type) {
    case CalendarType.MOTH:
      const monthData = getMonthFromState(data, state);
      currentComponent = (
        <CalendarMonth data={monthData} onClickEvent={onClickEvent} />
      );
      break;
    case CalendarType.WEEK:
      const weekData = getWeekFromState(data, state);
      currentComponent = (
        <CalendarWeek data={weekData} onClickEvent={onClickEvent} />
      );
      break;
    case CalendarType.DAY:
      const dayData = getDayFromState(data, state);
      currentComponent = (
        <CalendarDay data={dayData} onClickEvent={onClickEvent} />
      );
      break;
  }

  return (
    <div className="w-full">
      <div className="w-full flex flex-col">
        <CalendarNav
          currentView={type}
          onNav={onNav}
          onCreateEvent={onCreateEvent}
          onChangeView={onChangeView}
        />
        {currentComponent}
      </div>
    </div>
  );
};
