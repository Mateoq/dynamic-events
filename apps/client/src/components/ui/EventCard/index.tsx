import React from 'react';

import { Event } from '@/types';

export type EventCardProps = {
  data: Event;
  onClick?: (id: string) => void;
};

export const EventCard: React.FC<EventCardProps> = (props) => {
  const { data, onClick } = props;
  return (
    <button
      className="flex-1 px-1 overflow-hidden"
      style={{ background: data.color }}
      onClick={() => onClick?.(data.id)}
    >
      <p className="text-white text-sm truncate">{data.title}</p>
    </button>
  );
};
