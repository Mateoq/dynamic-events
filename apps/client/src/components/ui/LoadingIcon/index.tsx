import React from 'react';
import { motion } from 'framer-motion';
import { FaSpinner } from 'react-icons/fa6';

export type LoadingIconProps = {
  color?: string;
};

export const LoadingIcon: React.FC<LoadingIconProps> = (props) => {
  const { color = 'white' } = props;
  return (
    <motion.div
      className="w-4 h-4 round-full"
      animate={{ rotate: [0, 360] }}
      transition={{ duration: 1, ease: 'linear', repeat: Infinity }}
    >
      <FaSpinner color={color} />
    </motion.div>
  );
};
