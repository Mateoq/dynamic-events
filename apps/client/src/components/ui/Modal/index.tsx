import React from 'react';
import { AnimatePresence, motion } from 'framer-motion';
import { FaXmark } from 'react-icons/fa6';

export type ModalProps = {
  isOpen?: boolean;
  title: string;
  onClose: () => void;
};

export const Modal: React.FC<React.PropsWithChildren<ModalProps>> = (props) => {
  const { isOpen, title, children, onClose } = props;

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.div
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
          className="w-screen h-screen fixed top-0 left-0 flex items-center justify-center z-50 overflow-visible"
        >
          <div className="w-[500px] h-fit bg-white rounded shadow-2xl relative flex flex-col">
            <div className="flex justify-between h-fit py-2 px-2 bg-gray-500 rounded-t">
              <h3 className="text-lg text-gray-50 font-medium">{title}</h3>
              <button
                onClick={onClose}
                aria-label="Close Modal"
                className="h-6 w-6 transition-opacity hover:opacity-80 active:opacity-100"
              >
                <FaXmark size="100%" className="text-gray-50" />
              </button>
            </div>
            <div className="px-3 py-5">{children}</div>
          </div>
        </motion.div>
      )}
    </AnimatePresence>
  );
};
