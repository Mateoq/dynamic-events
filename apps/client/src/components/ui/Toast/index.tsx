import React from 'react';
import { AnimatePresence, motion } from 'framer-motion';
import clsx from 'clsx';

import { ToastType } from '@/types';

export type ToastProps = {
  isOpen?: boolean;
  type: ToastType;
  text: string;
};

export const Toast: React.FC<ToastProps> = (props) => {
  const { isOpen, type, text } = props;
  const isSuccess = type === ToastType.SUCCESS;
  const isError = type === ToastType.ERROR;

  return (
    <AnimatePresence>
      {isOpen && (
        <motion.div
          initial={{ x: '100%', opacity: 0 }}
          animate={{ x: 0, opacity: 1 }}
          exit={{ x: '100%', opacity: 0 }}
          transition={{ duration: 0.2 }}
          className={clsx(
            'w-52 min-h-24 rounded filter drop-shadow-lg flex items-center justify-center p-5 z-[100] fixed bottom-5 right-5',
            { 'bg-gray-900': isSuccess, 'bg-red-600': isError }
          )}
        >
          <span className="text-md text-gray-100">{text}</span>
        </motion.div>
      )}
    </AnimatePresence>
  );
};
