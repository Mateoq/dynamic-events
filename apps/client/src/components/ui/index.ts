export * from './Calendar';
export * from './LoadingIcon';
export * from './Modal';
export * from './Toast';
export * from './EventCard';
