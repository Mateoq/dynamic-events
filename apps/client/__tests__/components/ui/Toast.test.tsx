import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';

import { Toast } from '@/components/ui';
import { ToastType } from '@/types';
import { sleep } from '@/utils/functions';

describe('Toast', () => {
  it('should match snapshot', async () => {
    const { container } = render(
      <Toast isOpen type={ToastType.SUCCESS} text="Message" />
    );

    // Wait for fade in animation to be completed
    await sleep(400);
    const message = await screen.findByText('Message');

    expect(message).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
