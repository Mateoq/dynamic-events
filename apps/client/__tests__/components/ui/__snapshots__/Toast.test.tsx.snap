// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast should match snapshot 1`] = `
<div>
  <div
    class="w-52 min-h-24 rounded filter drop-shadow-lg flex items-center justify-center p-5 z-[100] fixed bottom-5 right-5 bg-gray-900"
    style="opacity: 1; will-change: transform; transform: none;"
  >
    <span
      class="text-md text-gray-100"
    >
      Message
    </span>
  </div>
</div>
`;
